File: src/main/java/org/jdc/template/ux/individualedit/IndividualEditPresenter.java
package org.jdc.template.ux.individualedit;

import com.google.android.gms.analytics.HitBuilders;

import org.jdc.template.Analytics;
import org.jdc.template.model.database.main.individual.Individual;
import org.jdc.template.model.database.main.individual.IndividualManager;
import org.jdc.template.ui.mvp.BasePresenter;
import org.threeten.bp.LocalDate;
import org.threeten.bp.LocalTime;

import javax.inject.Inject;

public class IndividualEditPresenter extends BasePresenter {
    private final IndividualManager individualManager;
    private final Analytics analytics;

    private IndividualEditContract.View view;
    private long individualId;
    private Individual individual = new Individual();

    @Inject
    public IndividualEditPresenter(IndividualManager individualManager, Analytics analytics) {
        this.individualManager = individualManager;
        this.analytics = analytics;
    }


    public void init(IndividualEditContract.View view, long individualId) {
        this.view = view;
        this.individualId = individualId;
    }

    @Override
    public void load() {
        loadIndividual();
    }

    private void loadIndividual() {
        individual = individualManager.findByRowId(individualId);
        if (individual != null) {
            analytics.send(new HitBuilders.EventBuilder()
                    .setCategory(Analytics.CATEGORY_INDIVIDUAL)
                    .setAction(Analytics.ACTION_EDIT)
                    .build());

            view.showIndividual(individual);
        }
    }

    public void alarmTimeClicked() {
        view.showAlarmTimeSelector(individual.getAlarmTime());
    }

    public void alarmTimeSelected(LocalTime time) {
            individual.setAlarmTime(time);
            view.showAlarmTime(time);
    }

    public void birthdayClicked() {
        view.showBirthDateSelector(individual.getBirthDate());
    }

    public void birthDateSelected(LocalDate date) {
            individual.setBirthDate(date);
            view.showBirthDate(date);
    }

    public void saveIndividual() {
        if (view.validateIndividualData()) {
            view.getIndividualDataFromUi(individual);

            individualManager.save(individual);
            analytics.send(new HitBuilders.EventBuilder().setCategory(Analytics.CATEGORY_INDIVIDUAL).setAction(Analytics.ACTION_EDIT_SAVE).build());

            view.close();
        }
    }
}

File: src/main/kotlin/org/jdc/template/ux/individualedit/IndividualEditPresenter.kt
package org.jdc.template.ux.individualedit

import com.google.android.gms.analytics.HitBuilders
import kotlinx.coroutines.experimental.CommonPool
import kotlinx.coroutines.experimental.Job
import kotlinx.coroutines.experimental.android.UI
import kotlinx.coroutines.experimental.launch
import kotlinx.coroutines.experimental.run
import org.jdc.template.Analytics
import org.jdc.template.model.database.main.individual.Individual
import org.jdc.template.model.database.main.individual.IndividualManager
import org.jdc.template.ui.mvp.BasePresenter
import org.threeten.bp.LocalDate
import org.threeten.bp.LocalTime
import javax.inject.Inject

class IndividualEditPresenter @Inject
constructor(private val analytics: Analytics,
            private val individualManager: IndividualManager): BasePresenter() {

    private lateinit var view: IndividualEditContract.View
    private var individualId = 0L
    private var individual: Individual? = null

    fun init(view: IndividualEditContract.View, individualId: Long) {
        this.view = view
        this.individualId = individualId
    }

    override fun load(): Job? {
        return loadIndividual()
    }

    private fun loadIndividual(): Job {
        val job = launch(UI) {
            if (individualId <= 0L) {
                individual = Individual()
                return@launch
            }

            individual = run(context + CommonPool) {
                individualManager.findByRowId(individualId)
            }

            individual?.let {
                analytics.send(HitBuilders.EventBuilder().setCategory(Analytics.CATEGORY_INDIVIDUAL).setAction(Analytics.ACTION_EDIT).build())
                view.showIndividual(it)
            }
        }
        compositeJob.add(job)
        return job
    }

    fun saveIndividual() {
        val job = launch(UI) {
            if (view.validateIndividualData()) {
                individual?.let {
                    view.getIndividualDataFromUi(it)

                    run(context + CommonPool) {
                        individualManager.save(individual)
                        analytics.send(HitBuilders.EventBuilder().setCategory(Analytics.CATEGORY_INDIVIDUAL).setAction(Analytics.ACTION_EDIT_SAVE).build())
                    }

                    view.close()
                }
            }
        }
        compositeJob.add(job)
    }

    fun alarmTimeClicked() {
        view.showAlarmTimeSelector(individual?.alarmTime ?: LocalTime.now())
    }

    fun alarmTimeSelected(time: LocalTime) {
        individual?.let {
            it.alarmTime = time
            view.showAlarmTime(time)
        }
    }

    fun birthdayClicked() {
        view.showBirthDateSelector(individual?.birthDate ?: LocalDate.now())
    }

    fun birthDateSelected(date: LocalDate) {
        individual?.let {
            it.birthDate = date
            view.showBirthDate(date)
        }
    }
}
